### ISSUE TITLE
Gracefully handle DBPedia returning a LANGSTRING without a language

### ISSUE THREAD

**Issue by reckart** [eclipse/rdf4j#2815]
Some SPARQL endpoints may not include language information on a LANGSTRING. The result parser currently fails in such situations. To gracefully handle this, the parser should downgrade the LANGSTRING to a simple STRING.

E.g. when querying DBPedia (http://de.dbpedia.org/sparql) for:

```
          SELECT DISTINCT ?lc ?subj
          WHERE { { { OPTIONAL { ?pLabel <http://www.w3.org/2000/01/rdf-schema#subPropertyOf>* <http://www.w3.org/2000/01/rdf-schema#label> . }
          { ?subj ?pLabel ?lc .
          ?lc <bif:contains> '''"lala"''' .
          FILTER ( ( ( REGEX( ?lc, "^.+lala$", "i" ) && LANGMATCHES( LANG( ?lc ), "de" ) ) || ( REGEX( ?lc, "^.+lala$", "i" ) && LANGMATCHES( LANG( ?lc ), "" ) ) ) ) } } } }
          LIMIT 6
```

[Execute query](http://de.dbpedia.org/sparql?default-graph-uri=&query=++++++++++SELECT+DISTINCT+%3Flc+%3Fsubj%0D%0A++++++++++WHERE+%7B+%7B+%7B+OPTIONAL+%7B+%3FpLabel+%3Chttp%3A%2F%2Fwww.w3.org%2F2000%2F01%2Frdf-schema%23subPropertyOf%3E*+%3Chttp%3A%2F%2Fwww.w3.org%2F2000%2F01%2Frdf-schema%23label%3E+.+%7D%0D%0A++++++++++%7B+%3Fsubj+%3FpLabel+%3Flc+.%0D%0A++++++++++%3Flc+%3Cbif%3Acontains%3E+%27%27%27%22lala%22%27%27%27+.%0D%0A++++++++++FILTER+%28+%28+%28+REGEX%28+%3Flc%2C+%22%5E.%2Blala%24%22%2C+%22i%22+%29+%26%26+LANGMATCHES%28+LANG%28+%3Flc+%29%2C+%22de%22+%29+%29+%7C%7C+%28+REGEX%28+%3Flc%2C+%22%5E.%2Blala%24%22%2C+%22i%22+%29+%26%26+LANGMATCHES%28+LANG%28+%3Flc+%29%2C+%22%22+%29+%29+%29+%29+%7D+%7D+%7D+%7D%0D%0A++++++++++LIMIT+6&should-sponge=&format=text%2Fhtml&timeout=0&debug=on)

Result:

| lc | subj
| -- | --
| `"Altin Lala"@de` | http://de.dbpedia.org/resource/Altin_Lala
| `"Hans Lala"@de` | http://de.dbpedia.org/resource/Hans_Lala
| `"Maharan Lala"@de` | http://de.dbpedia.org/resource/Maharan_Lala
| `"Jiri Lala"@de` | http://de.dbpedia.org/resource/Jiri_Lala
| `"Jan Lala"@de` | http://de.dbpedia.org/resource/Jan_Lala
| `"Hans Lala"^^<http://www.w3.org/1999/02/22-rdf-syntax-ns#langString>` | http://de.dbpedia.org/resource/Hans_Lala

```
Caused by: org.eclipse.rdf4j.query.QueryEvaluationException: org.eclipse.rdf4j.query.resultio.QueryResultParseException: java.lang.IllegalArgumentException: datatype rdf:langString requires a language tag
	at org.eclipse.rdf4j.query.impl.QueueCursor.convert(QueueCursor.java:58)
	at org.eclipse.rdf4j.query.impl.QueueCursor.convert(QueueCursor.java:22)
	at org.eclipse.rdf4j.common.iteration.QueueIteration.checkException(QueueIteration.java:165)
	at org.eclipse.rdf4j.common.iteration.QueueIteration.handleClose(QueueIteration.java:155)
	at org.eclipse.rdf4j.common.iteration.AbstractCloseableIteration.close(AbstractCloseableIteration.java:63)
	at org.eclipse.rdf4j.common.iteration.QueueIteration.checkException(QueueIteration.java:162)
	... 42 more
Caused by: org.eclipse.rdf4j.query.resultio.QueryResultParseException: java.lang.IllegalArgumentException: datatype rdf:langString requires a language tag
	at org.eclipse.rdf4j.query.resultio.sparqlxml.AbstractSPARQLXMLParser.parseQueryResultInternal(AbstractSPARQLXMLParser.java:208)
	at org.eclipse.rdf4j.query.resultio.sparqlxml.SPARQLResultsXMLParser.parseQueryResult(SPARQLResultsXMLParser.java:73)
	at org.eclipse.rdf4j.query.resultio.helpers.BackgroundTupleResult.run(BackgroundTupleResult.java:89)
	at org.eclipse.rdf4j.http.client.BackgroundResultExecutor.lambda$autoCloseRunnable$0(BackgroundResultExecutor.java:65)
	at java.base/java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:515)
	at java.base/java.util.concurrent.FutureTask.run(FutureTask.java:264)
	at java.base/java.util.concurrent.ScheduledThreadPoolExecutor$ScheduledFutureTask.run(ScheduledThreadPoolExecutor.java:304)
	at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1128)
	at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:628)
	at java.base/java.lang.Thread.run(Thread.java:834)
Caused by: java.lang.IllegalArgumentException: datatype rdf:langString requires a language tag
	at org.eclipse.rdf4j.model.impl.SimpleLiteral.<init>(SimpleLiteral.java:101)
	at org.eclipse.rdf4j.model.impl.SimpleValueFactory.createLiteral(SimpleValueFactory.java:132)
	at org.eclipse.rdf4j.query.resultio.sparqlxml.SPARQLResultsSAXParser.startTag(SPARQLResultsSAXParser.java:136)
	at org.eclipse.rdf4j.common.xml.SimpleSAXParser$SimpleSAXDefaultHandler.endElement(SimpleSAXParser.java:298)
	at java.xml/com.sun.org.apache.xerces.internal.parsers.AbstractSAXParser.endElement(AbstractSAXParser.java:610)
	at java.xml/com.sun.org.apache.xerces.internal.impl.XMLDocumentFragmentScannerImpl.scanEndElement(XMLDocumentFragmentScannerImpl.java:1718)
	at java.xml/com.sun.org.apache.xerces.internal.impl.XMLDocumentFragmentScannerImpl$FragmentContentDriver.next(XMLDocumentFragmentScannerImpl.java:2883)
	at java.xml/com.sun.org.apache.xerces.internal.impl.XMLDocumentScannerImpl.next(XMLDocumentScannerImpl.java:605)
	at java.xml/com.sun.org.apache.xerces.internal.impl.XMLNSDocumentScannerImpl.next(XMLNSDocumentScannerImpl.java:112)
	at java.xml/com.sun.org.apache.xerces.internal.impl.XMLDocumentFragmentScannerImpl.scanDocument(XMLDocumentFragmentScannerImpl.java:534)
	at java.xml/com.sun.org.apache.xerces.internal.parsers.XML11Configuration.parse(XML11Configuration.java:888)
	at java.xml/com.sun.org.apache.xerces.internal.parsers.XML11Configuration.parse(XML11Configuration.java:824)
	at java.xml/com.sun.org.apache.xerces.internal.parsers.XMLParser.parse(XMLParser.java:141)
	at java.xml/com.sun.org.apache.xerces.internal.parsers.AbstractSAXParser.parse(AbstractSAXParser.java:1216)
	at java.xml/com.sun.org.apache.xerces.internal.jaxp.SAXParserImpl$JAXPSAXParser.parse(SAXParserImpl.java:635)
	at org.eclipse.rdf4j.common.xml.SimpleSAXParser.parse(SimpleSAXParser.java:199)
	at org.eclipse.rdf4j.common.xml.SimpleSAXParser.parse(SimpleSAXParser.java:180)
	at org.eclipse.rdf4j.query.resultio.sparqlxml.AbstractSPARQLXMLParser.parseQueryResultInternal(AbstractSPARQLXMLParser.java:187)
	... 9 more
```

**Comment by hmottestad** [comment]
Just for info from the RDF 1.1 spec:

"if and only if the datatype IRI is http://www.w3.org/1999/02/22-rdf-syntax-ns#langString, a non-empty language tag as defined by [BCP47]. The language tag must be well-formed according to section 2.2.9 of [BCP47]." 

https://www.w3.org/TR/rdf11-concepts/#section-Graph-Literal

**Comment by reckart** [comment]
So this is the specific result that the parser has trouble with:

```
<sparql xmlns="http://www.w3.org/2005/sparql-results#" xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance" xsi:schemaLocation="http://www.w3.org/2001/sw/DataAccess/rf1/result2.xsd">
 <head>
  <variable name="lc"/>
  <variable name="subj"/>
 </head>
 <results distinct="false" ordered="true">
  <result>
   <binding name="lc"><literal xml:lang="de">Altin Lala</literal></binding>
   <binding name="subj"><uri>http://de.dbpedia.org/resource/Altin_Lala</uri></binding>
  </result>
  <result>
   <binding name="lc"><literal datatype="http://www.w3.org/1999/02/22-rdf-syntax-ns#langString">Hans Lala</literal></binding>
   <binding name="subj"><uri>http://de.dbpedia.org/resource/Hans_Lala</uri></binding>
  </result>
 </results>
</sparql>
```

The literal with the language is parsed ok, but not the one with the datatype.

I am building a PR with a proposed fix and a unit test - will be ready in a jiffy. Let's discuss it then.

**Comment by reckart** [comment]
Here it is. I understand that this is invalid according to the RDF spec - but this puts users in between a rock and a hard place:

- one can choose to use RDF4J but one cannot communicate with some popular SPARQL endpoints (i.e. DBPedia)
- one can choose to contact the people who run the SPARQL endpoint and have them fix the problem - but one still cannot use the endpoint until they do it (if they ever do it)

... one can choose not to use RDF4J - or to fork it and apply the fix in the fork... that is actually what we did in INCEpTION, but we want to get away from that and use the stock releases - in particular now that newer versions of Lucene/Solr/Elasticsearch are supported. But there still is this problem here with SPARQL endpoints that are not fully standards compliant...

**Comment by reckart** [comment]
Btw. I also opened an issue with DBPedia: https://github.com/dbpedia/extraction-framework/issues/676

**Comment by hmottestad** [comment]
I think @jeenbroekstra is more qualified for this discussion than me.

**Comment by jeenbroekstra** [comment]
Related issue: #2004

**Comment by jeenbroekstra** [comment]
Thanks for raising this @reckart , and thanks for the PR as well. 

For the RDF syntax parsers, this behavior is controlled by a configuration flag: if `BasicParserSettings.VERIFY_DATATYPE_VALUES` (which is by default set to `true`) is changed to `false`, it accepts rdf:langString literals without an actual language tag. This is done this way because our parsers by default are strict, to allow them to be used for validation purposes. 

For the SPARQL query result formats, I don't think we need to follow that practice however: I think doing syntax validation for query results is something that rarely if ever is done by users, so I'm fine with these parsers being a bit more lenient by default.

As I also mentioned on the PR: the fix you provided addresses this for the SPARQL/XML result format, but not for SPARQL/JSON. Would you be able to add the same fix for that parser as well to your PR, so we fix it in  both on one ticket?

**Comment by reckart** [comment]
@jeenbroekstra I'll take a stab at it. Is there a particular deadline this needs to meet in order to make it into 3.6.0?

**Comment by jeenbroekstra** [comment]
> @jeenbroekstra I'll take a stab at it. Is there a particular deadline this needs to meet in order to make it into 3.6.0?

Thanks! The 3.6 release is officially overdue (it was originally planned for end of December). @hmottestad is taking point on it, as his rewrite of the SHACL engine is the Big New Feature for this release. That dictates the timeline. There's a milestone build done just yesterday, but I believe there are one or two bugs that he still has to spend some time on. 

The short of it is that I don't know exactly when we'll be ready to release, but I expect that it won't be in the next week or so.

**Comment by hmottestad** [comment]
I'm not certain myself about the timeframe. All I can say is it's going to be days to weeks. Had hoped to have it done last weekend.

**Comment by jeenbroekstra** [comment]
Thanks again @reckart , this is now merged and will be in the upcoming 3.6 release!
